pkg/resources/kafka/kafka.go

```diff
@@ -232,17 +240,37 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 
 	var brokerPods corev1.PodList
 	matchingLabels := client.MatchingLabels(apiutil.LabelsForKafka(r.KafkaCluster.Name))
-	err = r.Client.List(context.TODO(), &brokerPods, client.ListOption(client.InNamespace(r.KafkaCluster.Namespace)), client.ListOption(matchingLabels))
+	err = r.Client.List(ctx, &brokerPods, client.ListOption(client.InNamespace(r.KafkaCluster.Namespace)), client.ListOption(matchingLabels))
 	if err != nil {
 		return errors.WrapIf(err, "failed to list broker pods that belong to Kafka cluster")
 	}
 
+	runningBrokers := make(map[string]struct{})
+	for _, b := range brokerPods.Items {
+		brokerID := b.GetLabels()["brokerId"]
+		runningBrokers[brokerID] = struct{}{}
+	}
+
 	controllerID, err := r.determineControllerId()
 	if err != nil {
 		log.Error(err, "could not find controller broker")
 	}
 
-	reorderedBrokers := reorderBrokers(brokerPods, r.KafkaCluster.Spec.Brokers, r.KafkaCluster.Status.BrokersState, controllerID)
+	var pvcList corev1.PersistentVolumeClaimList
+	err = r.Client.List(ctx, &pvcList, client.ListOption(client.InNamespace(r.KafkaCluster.Namespace)), client.ListOption(matchingLabels))
+	if err != nil {
+		return errors.WrapIf(err, "failed to list broker pvcs that belong to Kafka cluster")
+	}
+
+	boundPersistentVolumeClaims := make(map[string]struct{})
+	for _, pvc := range pvcList.Items {
+		brokerID := pvc.GetLabels()["brokerId"]
+		if pvc.Status.Phase == corev1.ClaimBound {
```

Comment:
Bound PVC means that there is a PV that matches the PVC and is bound to the PVC: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#binding

A PV can be created without being attached to a VM (Kubernetes node). e.g. if you create a PVC which uses a Storage class that has `volumeBindingMode: Immediate` a PV will be provisioned and bound to the PVC without the disk being mounted to any VM. The spec of the PV provisioned dynamically through the PVC will indicate which zone it was provisioned in:
```
apiVersion: v1
kind: PersistentVolume
...
spec:
  ...
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.kubernetes.io/zone
          operator: In
          values:
          - us-east-2a
        - key: topology.kubernetes.io/region
          operator: In
          values:
          - us-east-2
  persistentVolumeReclaimPolicy: Delete
  storageClassName: gp2immediate
  volumeMode: Filesystem
status:
  phase: Bound
```
(see https://kubernetes.io/docs/concepts/storage/persistent-volumes/#node-affinity)

If PV will be actually attached to a VM when it its mounted to a Pod in which case it will be attached to the VM that hosts the Pod.
If a Storage class with `volumeBindingMode: WaitForFirstConsumer` is used the PV won't be provisioned until there is no Pod to use it. When the scheduler decided which node to schedule the Pod onto the PV will be provisioned in the zone of the Pod.

When the VM dies and the PV is not lost, the Pod on the VM vanishes. When the missing pods a recreated if those can be recreated on another node (VM) in the same zone Kubernetes will reattach the disk to the other node(VM) . In case the scheduler cannot schedule the pods to another in the same zone but in other zone for various reasons the pods will stuck in pending state waiting for volume mounting which will never happen as disks can be attached only to VMs in same zone.

When the VM dies and the PV is lost as well the PVC will get into `phase: Lost` which indicates the underlying PV does not exist any longer and all data on it was lost.

There might be some differences from one cloud provider to another thus always worth checking their docs regarding the volume provisioning plugins/drivers they provide for Kubernetes but I think we can rely on the `phase: Bound` field of the PVC regardless of the env as that indicates if there is an existing PV linked to the PVC. The reattaching of the disk to another VM is a different story and is environment specific how exactly works and impacts the Pod scheduling.

Reply:
Thanks for the detailed explanation, makes sense